docs/dependency-management.md

```diff
@@ -0,0 +1,50 @@
+# Dependency Management Guidelines
+
+This repo uses a number of third-party rust crates, so we need to be mindful
+of how these dependencies will affect our consumers.  Considerations include:
+
+* General code quality
+* [Licensing compatibility](https://www.mozilla.org/en-US/MPL/license-policy/#Licenses_Compatible_with_the_MPL)
+* Handling of security vulnerabilities
+* The potential for [supply-chain compromise](https://medium.com/intrinsic/compromised-npm-package-event-stream-d47d08605502)
+
+We're still evolving our policies in this area, but here are the
+guidelines we've developed so far:
+
+* When adding a new dependency:
+  * In general, be conservative in adding new third-party dependencies.
+    * For trivial functionality, consider just writing it yourself.
+      Remember the cautionary tale of [left-pad](https://www.theregister.co.uk/2016/03/23/npm_left_pad_chaos/).
+    * Check if we already have a crate in our dependency tree that can provide the needed functionality.
+    * Prefer widely-used crates or crates from rust-lang-nursery.
+  * Check that it is clearly licensed and is [MPL-2.0 compatible](https://www.mozilla.org/en-US/MPL/license-policy/#Licenses_Compatible_with_the_MPL).
+  * Take the time to investigate the crate's source and ensure it is suitably high-quality.
+    * Be especially wary of uses of `unsafe`.
+    * Use `cargo fetch` to obtain the same source code that cargo will eventually build.
+  * Transitively apply this reasoning to all of its dependencies.
+  * Always depend on a specific version of the dependency, avoiding semver pattern-matching.
```

Comment:
I'm not confident this is viable, it will just lead to build errors and duplication. I think the closest we can come to this is being deliberate about updates to Cargo.lock.

That or vendoring.

Note that "a.b.c" in Cargo.toml is a semver match for 'compatible with "a.b.c"'

Reply:
> Note that "a.b.c" in Cargo.toml is a semver match for 'compatible with "a.b.c"'

Hrm, to be sure I understand: does this mean that if you specify "1.2.3" in your Cargo.toml dependencies, but some other dependency requires "1.2.4", then the resulting Cargo.lock will contain just "1.2.4"?

Reply:
Yes. Additionally, if 1.2.5 exists when the Cargo.lock file is generated, it will contain that and not 1.2.3 or 1.2.4.

Reply:
And while we can control this for ourselves, we can't for dependencies. We're very likely to cause cargo to give up in disgust if we try to enforce this strictly, it doesn't have good support for that, the intent is if you want this you use a lockfile.

Reply:
This sums it up well as ["defaults to caret semantics"](https://stackoverflow.com/a/45224646).